### DISCUSSION TITLE
Add a ability to check error types in the middleware

### DISCUSSION THREAD

**Discussion by vchirikov** [withastro/roadmap#950] [Proposal]
### Body

# Summary

At the moment, Astro does not give the opportunity to use Error Types in the middlewares. Let's export types & helpers to the users. Moreover Astro shouldn't use the same error type to general rewrite pipeline error and 404.

# Background & Motivation

Let's imagine I want to write custom i18n middleware which should rewrite to the same page with detected locale in `Astro.locals`.

```ts
import type { MiddlewareHandler } from 'astro';

import { isAstroError } from 'astro:core/errors';
// This function isn't exported yet ^^
import { RewriteEncounteredAnError } from 'astro:core/errors/errors-data';
// The error types aren't exported too ^^

import { defaultLocale, extractLocale } from '@/lib/i18n';
import { removeTrailingSlash } from '@/lib/utils/url';

const frontendUrl = removeTrailingSlash(process.env.FRONTEND_URL);
const frontendUrlLen = frontendUrl.length;

export const onRequest: MiddlewareHandler = async (ctx, next) => {
  // ctx.currentLocale is broken, e.g. with base /base/ it found locale inside url /base/foo/en/bar
  // it also can't be set via middleware, so use next(pageUrl) & locals
  if (ctx.request.url.startsWith(frontendUrl)) {
    const urlWithoutBase = ctx.request.url.slice(frontendUrlLen);
    const { locale, url } = extractLocale(urlWithoutBase);
    ctx.locals.locale = locale ?? defaultLocale;
    try {
      await next(url);
    }
    catch (error: unknown) {
      // in case of 404/rendering error astro will throw the SAME error type :(
      if (isAstroError(error) && error.name == RewriteEncounteredAnError.name) {
        // but I want to silent only rewrites to non-existed pages (404),
        // Here I can't differentiate it's wrong route from user or rendering error, 
        // but, I think the middleware should know what kind of error were in pipeline execution
        // I also can't rewrite after page not found to 404 because "Rewriting a 404 is only allowed inside on-demand pages.":
        await next('/404');
        // ^^^ doesn't work in SSR too T_T
        return;
      }
    }
  }
  return await next();
};

```

Now this code doesn't work, because `astro/error` only export `AstroError`.

https://github.com/withastro/astro/blob/2da877bfe97144ff5110d011b9bfba270ccfc620/packages/astro/package.json#L78

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- pipeline is more flexible (e.g. for i18n)
- users might use internal types from `core/errors`
- users can rewrite into `/404` inside pipeline execution (with or without context dropping ctx.next('/404') or next(`/404`) should be available)

**Comment by millette** [comment]
The [public roadmap](https://github.com/orgs/withastro/projects/11/views/1) links to #951 `https://github.com/withastro/roadmap/issues/951` (which is a 404) instead of here, not sure how that happened.

**Comment by millette** [comment]
Could this be used to retrieve the Response status code? If a page

```js
returns new Response(null, { status: 403 })
```

I would like to use that info in the middleware.
